File: dan/Problems/Easy/LinkedList/160. Intersection of Two Linked Lists/solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        a, b = headA, headB
        dic = {}
        while a:
            dic[a] = 1
            a = a.next
        while b:
            if b in dic:
                return b
            else:
                b = b.next
        return None
    
    def getIntersectionNode(self, headA, headB):
        a, b, flag = headA, headB, 2
        while a and b and flag:
            if a == b:
                return a
            if not a.next:
                a = headB
                flag -= 1
            else:
                a = a.next
            if not b.next:
                b = headA
            else:
                b = b.next
            
        return None

File: dan/Problems/Easy/LinkedList/160. Intersection of Two Linked Lists/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    //my solution
    public ListNode getIntersectionNode1(ListNode headA, ListNode headB) {
        if(headA == null || headB == null)
            return null;
            
        ListNode cur1 = headA;
        ListNode cur2 = headB;
        int sup = 0;
        while(sup < 3){
            if(cur1 == cur2)
                return cur1;
            
            if(cur1.next == null){
                cur1 = headB;
                sup++;
            }else
                cur1 = cur1.next;
                
            if(cur2.next == null){
                cur2 = headA;
                sup++;
            }else
                cur2 = cur2.next;
                    
        }
        return null;
    }
    
    //smarter solution: if no intersection, cur1 and cur2 will meet with both the values are null.
    public ListNode getIntersectionNode(ListNode headA, ListNode headB){
        ListNode curA = headA;
        ListNode curB = headB;
        while(curA != curB){
            curA = curA==null?headB:curA.next;
            curB = curB==null?headA:curB.next;
        }
        return curA;
    }
}
